src/PaprController.sol

```diff
@@ -261,35 +275,55 @@ contract PaprController is
+            args.auction.auctionAssetContract, ReservoirOracleUnderwriter.PriceKind.TWAP, args.oracleInfo
         ) * count;
+        bool isLastCollateral = count == 0;
 
-        uint256 debtCached = _vaultInfo[auction.nftOwner][auction.auctionAssetContract].debt;
-        uint256 maxDebtCached = count == 0 ? 0 : _maxDebt(collateralValueCached, updateTarget());
+        uint256 debtCached = _vaultInfo[args.auction.nftOwner][args.auction.auctionAssetContract].debt;
+        uint256 maxDebtCached = isLastCollateral ? 0 : _maxDebt(collateralValueCached, updateTarget());
         /// anything above what is needed to bring this vault under maxDebt is considered excess
         uint256 neededToSaveVault = maxDebtCached > debtCached ? 0 : debtCached - maxDebtCached;
-        uint256 price = _purchaseNFTAndUpdateVaultIfNeeded(auction, maxPrice, sendTo);
+        uint256 price = _purchaseNFTAndUpdateVaultIfNeeded(args.auction, args.maxPrice, args.sendTo);
         uint256 excess = price > neededToSaveVault ? price - neededToSaveVault : 0;
         uint256 remaining;
+        uint256 newDebtCached;
 
         if (excess > 0) {
-            remaining = _handleExcess(excess, neededToSaveVault, debtCached, auction);
+            uint256 fee = excess * liquidationPenaltyBips / BIPS_ONE;
+            uint256 credit = excess - fee;
+            uint256 totalOwed = credit + neededToSaveVault;
+
+            PaprToken(address(papr)).burn(address(this), fee);
+
+            if (totalOwed > debtCached) {
+                // we owe them more papr than they have in debt
+                // so we pay down debt and send them the rest
+                newDebtCached = _reduceDebt(
+                    args.auction.nftOwner, args.auction.auctionAssetContract, address(this), debtCached, debtCached
+                );
+                papr.transfer(args.auction.nftOwner, totalOwed - debtCached);
+            } else {
+                // reduce vault debt
+                newDebtCached = _reduceDebt(
+                    args.auction.nftOwner, args.auction.auctionAssetContract, address(this), debtCached, totalOwed
+                );
+                remaining = debtCached - totalOwed;
+            }
         } else {
-            _reduceDebt(auction.nftOwner, auction.auctionAssetContract, address(this), price);
+            newDebtCached =
+                _reduceDebt(args.auction.nftOwner, args.auction.auctionAssetContract, address(this), debtCached, price);
             remaining = debtCached - price;
         }
 
-        if (count == 0 && remaining != 0) {
+        if (isLastCollateral && remaining != 0) {
             /// there will be debt left with no NFTs, set it to 0
-            _reduceDebtWithoutBurn(auction.nftOwner, auction.auctionAssetContract, remaining);
+            _reduceDebtWithoutBurn(args.auction.nftOwner, args.auction.auctionAssetContract, newDebtCached, remaining);
```

Comment:
maybe I don't need `newDebtCached` and this could just be `remaining`, as well? Should be same if we've done it right